frontend/src/App.js

```diff
@@ -2,9 +2,45 @@ import React, { Component } from 'react';
 import './App.css';
 
 class App extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      quote: 'They don\'t know that we know they know we know.',
+      typedText: ''
+    };
+    this.handleChange = this.handleChange.bind(this);
+  }
+
+  handleChange(event) {
+    const typedText = event.target.value;
+    this.setState({ typedText: typedText });
+  }
+
+  getDisplayQuote() {
+    const quoteArray = this.state.quote.split('');
+    const typedArray = this.state.typedText.split('');
+    return (
+      quoteArray.map((letter, index) => {
+        if (!(index in typedArray)) {
```

Comment:
An iterable function worth looking at here is `zip`, which isn't built into JS, but works like so:

```javascript
zip([x1, x2, x3], [y1, y2, y2]) === [[x1, y1], [x2, y2], [x3, y3]]
```

It takes two lists and "zips" them together into pairs.  If the lists are a different size, you can make it so the resulting array is as big as the bigger list and the missing spots are filled with `undefined`.

Lodash has the method: https://lodash.com/docs/4.17.11#zip

So, you'd get something like:

```javascript
quoteArray.zip(typedArray).map((quoteChar, typedChar) => {
  if (typedChar === undefined) {
    return <span className="untyped-letter">{letter}</span>
  } else if (typedChar !== quoteChar) {
    ...
  } else {
    ...
  }
});
```

Or, there's a common variation called `zipWith` that combines the zip and the map.  In Lodash: https://lodash.com/docs/4.17.11#zipWith

This works like:

```javascript
zipWith(fn, [x1, x2, x3], [y1, y2, y2]) === [fn(x1, x2), fn(x2, y2), fn(x3, y3)]
```

So your code is then:

```javascript
quoteArray.zipWith(typedArray, (quoteChar, typedChar) => {
  // same function body as above
});
```

You'd have to implement `zip` or `zipWith` yourself (or use a library like lodash).

This removes all the index-related bookkeeping in your code.

Reply:
Not saying you need to make this change, just pointing out that the pattern you're using is more naturally expressed via `zipWith`.

Reply:
alternatively a way to go about this might be to use some kind of substring matching